### PR TITLE
Utility function to check if numbers in one string exists in another 

### PR DESCRIPTION
**Pull request by Eyobyb**
Number exisitance checking utility

This pull request introduces a utility function designed to determine whether the "source" set of text data contains the numbers referenced in the "result" set.

This utility function will serve as a temporary solution for verifying whether the numerical results from the language model originated from scraped data rather than being hallucinated.

### CODE REVIEW

File: src/sherpa_ai/utils.py [hunk 1]
```diff
@@ -300,3 +304,19 @@ def check_url(url):
...
+    check_numbers = extract_numbers_from_text(result)
+    source_numbers = extract_numbers_from_text(source)
+    source_link =  source_link if source_link is not None else "source data"
+    message = []
+    for data in check_numbers:
+        if data not in source_numbers:
+            message.append(f"{data} is not mentioned in the {source_link}. ")
+    if len(message)>0:
+        return {"number_exisit": False , "messages":message}
+    return {"number_exisit": True , "messages":message}
```
Comment (20001LastOrder) [inline src/sherpa_ai/utils.py:322]: Please add new lines at the end of files

File: src/tests/unit_tests/test_util.py [hunk 2]
```diff
@@ -133,4 +135,16 @@ def test_log_formatter_formats_correctly_2(logs_with_final_response):
...
+    check_result = check_if_number_exist(source ,result , 'jack.com')
+    assert check_result['number_exisit'] == True
+
+def test_extract_numbers_from_text_fails():
+    source = " Cillum labore et culpa elit irure labore nostrud 12.45 minim cupidatat. Nulla nisi aliquip do duis elit tempor magna. Occaecat sunt nisi aliqua officia fugiat. Dolor ea ad mollit nulla ullamco sit voluptate cillum id laboris et proident anim. Culpa officia incididunt sit qui exercitation magna voluptate Lorem duis eu occaecat. Non occaecat deserunt voluptate cillum aliquip voluptate veniam. Ullamco commodo eiusmod consequat dolor cillum quis Lorem $45,000 labore tempor cupidatat  7 elit quis deserunt.  "
+    result = "Labore deserunt 12.45 $45,000  sit velit nulla. Sint ipsum reprehenderit sint cupidatat amet est id anim exercitation fugiat adipisicing elit. Id est dolore minim magna occaecat aute. Est dolore culpa laborum non esse nostrud."
+    check_result = check_if_number_exist(source ,result , 'jack.com')
+    assert check_result['number_exisit'] == False
```
Comment (20001LastOrder) [inline src/tests/unit_tests/test_util.py:150]: Besides asserting the existence, we also want to check the numbers extracted are correct...

**Review by 20001LastOrder — CHANGES_REQUESTED** [review]
Thanks for the PR, please check the comments below.

File: src/sherpa_ai/utils.py [hunk 3]
```diff
@@ -300,3 +304,19 @@ def check_url(url):
...
+def check_if_number_exist(result:str, source:str , source_link:str or None = None):
+    check_numbers = extract_numbers_from_text(result)
+    source_numbers = extract_numbers_from_text(source)
+    source_link =  source_link if source_link is not None else "source data"
+    message = []
+    for data in check_numbers:
+        if data not in source_numbers:
+            message.append(f"{data} is not mentioned in the {source_link}. ")
+    if len(message)>0:
+        return {"number_exisit": False , "messages":message}
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:321]: typo number_exisit => number_exists

File: src/sherpa_ai/utils.py [hunk 4]
```diff
@@ -300,3 +304,19 @@ def check_url(url):
...
     
+def extract_numbers_from_text(text):
+    pattern = r"\d+\.\d+|\d+\,\d+|\d+"
+    matches = re.findall(pattern, text)
+
+    return matches
+def check_if_number_exist(result:str, source:str , source_link:str or None = None):
+    check_numbers = extract_numbers_from_text(result)
+    source_numbers = extract_numbers_from_text(source)
+    source_link =  source_link if source_link is not None else "source data"
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:315]: source_ink = source_link or "source data"

File: src/sherpa_ai/utils.py [hunk 5]
```diff
@@ -300,3 +304,19 @@ def check_url(url):
     else:
         return True
     
+def extract_numbers_from_text(text):
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:307]: Add unit tests to assure yourself this regex pattern works against the types of numbers you are expecting to match

File: src/sherpa_ai/utils.py [hunk 6]
```diff
@@ -300,3 +304,19 @@ def check_url(url):
...
+    pattern = r"\d+\.\d+|\d+\,\d+|\d+"
+    matches = re.findall(pattern, text)
+
+    return matches
```
Comment (oshoma) [inline src/sherpa_ai/utils.py:311]: add newline after `return matches` and delete newline before

**Review by oshoma — CHANGES_REQUESTED** [review]

File: src/tests/unit_tests/test_util.py (hunk 2 above)
Comment (Eyobyb) [reply src/tests/unit_tests/test_util.py]: I added a separate test specifically for that.

**Pull request closed without merging**